metrics/ci_integrations/test/integration/ci/ci_integration_test.dart

```diff
@@ -2,242 +2,115 @@
-      () async {
-        final sourceClient = SourceClientStub(
-          fetchBuildsAfterCallback: (_, __) => throw UnimplementedError(),
-        );
-        final ciIntegration = CiIntegrationStub(sourceClient: sourceClient);
-        final result = await ciIntegration.sync(syncConfig);
-
-        expect(result.isError, isTrue);
+        verifyInOrder([
+          firstSyncStage.call(syncConfig),
+          secondSyncStage.call(syncConfig),
+        ]);
       },
     );
 
     test(
-      ".sync() returns an error if a destination client throws fetching the last build",
+      ".sync() does not continue the sync if one of the stages fails",
       () async {
-        final destinationClient = DestinationClientStub(
-          fetchLastBuildCallback: (_) => throw UnimplementedError(),
-        );
-        final ciIntegration = CiIntegrationStub(
-          destinationClient: destinationClient,
-        );
-        final result = await ciIntegration.sync(syncConfig);
+        when(firstSyncStage.call(syncConfig)).thenErrorWith();
 
-        expect(result.isError, isTrue);
+        await ciIntegration.sync(syncConfig);
+
+        verifyNever(secondSyncStage.call(any));
       },
     );
 
     test(
-      ".sync() returns an error if a destination client throws adding new builds",
+      ".sync() returns an error if one of the stages fails",
       () async {
-        final destinationClient = DestinationClientStub(
-          addBuildsCallback: (_, __) => throw UnimplementedError(),
-        );
-        final ciIntegration = CiIntegrationStub(
-          destinationClient: destinationClient,
-        );
+        when(firstSyncStage.call(syncConfig)).thenErrorWith();
+
         final result = await ciIntegration.sync(syncConfig);
 
         expect(result.isError, isTrue);
       },
     );
 
     test(
-      ".sync() ignores empty list of new builds and not call adding builds",
+      ".sync() returns an error with the message from the failed sync stage",
       () async {
-        final sourceClient = SourceClientStub(
-          fetchBuildsAfterCallback: (_, __) => Future.value([]),
-        );
-        final destinationClient = DestinationClientStub(
-          addBuildsCallback: (_, __) => throw UnimplementedError(),
-        );
-        final ciIntegration = CiIntegration(
-          sourceClient: sourceClient,
-          destinationClient: destinationClient,
-        );
-        final result = await ciIntegration.sync(syncConfig);
+        const expectedMessage = 'test';
+        when(
+          firstSyncStage.call(syncConfig),
+        ).thenErrorWith(null, expectedMessage);
 
-        expect(result.isSuccess, isTrue);
-      },
-    );
-
-    test(
-      ".sync() synchronizes builds",
-      () async {
-        final ciIntegration = CiIntegrationStub();
         final result = await ciIntegration.sync(syncConfig);
 
-        expect(result.isSuccess, isTrue);
+        expect(result.message, contains(expectedMessage));
```

Comment:
Maybe an algorithm should return the same result if this result is an error? So there won't be confusing messages like 'Failed to sync: Failed due to error:..`? Otherwise, stages should error with appropriate messages.